### ISSUE TITLE
Automatic language detection plan

### ISSUE THREAD

**Issue by TylerLeonhardt** [microsoft/vscode#129004]
Continuing from https://github.com/microsoft/vscode/issues/118455 ... this will span multiple milestones.

### Where we are

Since [this PR](https://github.com/microsoft/vscode/pull/128708) is merged in which brings in [vscode-languagedetection](https://github.com/Microsoft/vscode-languagedetection), we now have automatic language detection. What's it all about?
* 100% no code leaves your VS Code instance. Model is queried locally.
* opt-in feature
   * `"workbench.editor.untitled.languageDetection": true`
   * supports language specific enablement: `"[plaintext]" { "workbench.editor.untitled.languageDetection": true }`
* Powered by @yoeo's [guesslang](https://github.com/yoeo/guesslang) model (the latest release of it) which supports 30 languages
* Very basic additional heuristics to help the model with accuracy: (adding confidence of JS and TS, and C and C++
* Uncompressed (~4MB package)

This provides an "ok" experience but the model has to be very very sure it's the language it thinks it is to get the untitled file to change. You can enable the feature, and paste in a pretty large sample of code, and it should work.

### Where we wanna be

* Support as many languages as possible (JSON, YAML, XML, are not supported today for example)
* You should be able to open an untitled file and start typing and the language detection flips on as fast as possible
* A nice experience to handle tie breakers, "almost confident but not confident enough" situations and when we are wrong
* Compressed as much as possible (preliminary tests seem to say we can get down to 2-3MB)

### How we'll get there

* Improved guesslang model
  * [x] https://github.com/yoeo/guesslang/pull/33 which adds support for 14 more languages (JSON, YAML, XML included)
  * [ ] Possibly help guesslang with more files to train on
* Improved heuristics
  * [x] Variable confidence acceptance (i.e. If the model is 30% confident it's Java, but <1% confident it's anything else, then it's probably Java and we should set that)
  * [ ] https://github.com/microsoft/vscode/issues/129596
  * [x] ~Look at the user's workspace and influence weight based on what's open~ sounds too costly of an operation
* Improved UX
  * [x] ~In the event of a tie breaker, show a notification or similar to say "I'm tied between these X languages. which one is it?"~
  * [x] Make sure the user's decision doesn't get overwritten
  * [x] If we were wrong, promote the language picker (maybe show a badge on the language status bar entry when we change the lang)
  * [x] add detected languages to the top of the language picker
* Improved Perf
  * [ ] Compress the model as much as possible
  * [x] Handle large files (i.e. someone pasting a huge JSON payload)
  * [x] debounce event for untitled files
* Improved feedback
  * [x] It's important to understand how the model is doing for users to make sure it's actually useful. To do that, we opened https://github.com/microsoft/vscode/issues/129576

### Additional possible investigations

* [x] Have `code -` open untitled files and detect the language
* Kernel guessing of a Jupyter Notebook
* Provide feedback to the model so that it can learn from users? (this would be totally local)
* Use a different model than guesslang

**Comment by 4086606** [comment]
Maybe `I'm tied between these X languages. which one is it?` should sort language entries by their popularity inside the repository.

**Comment by 4086606** [comment]
I also think VSC should allow differentiating between different file associations such as MATLAB and Objective-C automatically. There's prior art in this area over at https://github.com/neel1996/langline/tree/432f964

**Comment by JacksonKearl** [comment]
> Variable confidence acceptance

We do something vaguely like this in the triage bot. Basically:
1. Train the model on some data
2. Using never-before-seen data, generate labels for the data and keep track of the model's raw output value for each labeling
3. For each label, go through all the raw output values for that label and find the cutoff where say 80% of labelings with a raw value above that cutoff were in fact that label. For example if you had data like:

| type | raw_js_model_output |
|---|---|
| js | 4.4 |
| js | 3.6 |
| js | 3.5 |
| ts | 3.4 |
| js | 3.2 |
| ts | 3.0 |

you would set the cutoff to be 3.2 for 80% confidence.

I haven't actually seen this approach documented anywhere, but it works well in the issue triaging bot.

**Comment by TylerLeonhardt** [comment]
@JacksonKearl is your suggestion more of something that should be done in the model itself? Or is this something I can get TFJS to give me?

**Comment by JacksonKearl** [comment]
I'd say the consumers of the model (us!) should be responsible for final calibrations of the applications confidence thresholds based on performing the above algorithm with data representative of what will actually be passed to the model in the target scenario. 

IMO the most important next step (which should unblock lots of other steps) is finding data representative of what will be actually passed to the model. This would ideally include lots of small/incomplete snippets similar to what one might write when just starting out in a new untitled buffer as opposed to complete files as-checked-in to the main branches of various OSS repos... I do not know a good source for such data, but it'd be invaluable.

As stands, we can't really make any useful claims regarding the accuracy of the detector. I know the guesslang project identifies with >90% accuracy full complete files as checked in to main branches of SCM, which is great, but that's not really what we're working with.

**Comment by TylerLeonhardt** [comment]
So it sounds like you're more speaking to the point of the possible investigations:

> Use a different model than guesslang

Just trying to classify your comment.

**Comment by JacksonKearl** [comment]
The act of threshold configuring is orthogonal to the model choice, the idea in threshold configuration is to take your real data and develop and understanding of how a given model responds to it, and from that form a way to look at the raw output of the model on a new piece of data and decide if you should be confident in that labeling or not. This isn't something guesslang can do themselves because they don't have the same type of input data as we will be providing. 

Going through this process could potentially solve 
> Use what we know about VS Code users to weigh languages that are most likely to come up (TS over coffeescript)

as well. If the validation data we provide generally matches the distribution of vscode's language usage, we might get this for free.

**Comment by 4086606** [comment]
Does https://github.com/microsoft/vscode/issues/129004#issuecomment-882751887 fall under the scope of this issue? Or should I open a new issue

**Comment by TylerLeonhardt** [comment]
@4086606 hmmm I feel like that should be a new issue. It theoretically could use the same model though. That said, with the exception of MATLAB and Objective-C (which... Don't have the largest representation in vscode) are there other examples of file association conflicts?

**Comment by 4086606** [comment]
https://github.com/github/linguist/blob/617fa486aad61043996e1323a429c900493c89a7/lib/linguist/heuristics.yml#L30 but it's alphabetically sorted

**Comment by TylerLeonhardt** [comment]
I think that should be a new issue. I also think there are different heuristics that would be better like looking at what other files are in the workspace and see what their modes are.

**Comment by TylerLeonhardt** [comment]
In tomorrow's insiders, language detection will now honor the language picker. It says `(Auto)` while Untitled Language Detection is on but then once you pick the language picker and manually select a language, the auto goes away and you will not be switched off from that language.
![image](https://user-images.githubusercontent.com/2644648/126575061-deb46a4a-85cf-49c2-a300-8b0a8326f0a5.png)

**Comment by 4086606** [comment]
I do agree that the regexes are ugly but if they scale to GitHub.. new issue coming up

**Comment by TylerLeonhardt** [comment]
The language picker now shows detected languages at the top.
![image](https://user-images.githubusercontent.com/2644648/126711571-c220fe8f-b69d-447d-b1fd-506bf6942435.png)

With that experience I don't think we need this anymore:

> In the event of a tie breaker, show a notification or similar to say "I'm tied between these X languages. which one is it?"

In my testing, the model did fairly well at detecting which language was which. It's very rare that it was a tie, even for languages like TS & JS and C & C++ which is interesting.

> Look at the user's workspace and influence weight based on what's open

This sounds costly to me so I think I'm going to skip it for now.

**Comment by TylerLeonhardt** [comment]
btw, I've changed the setting to `workbench.editor.untitled.experimentalLanguageDetection` to be explicit in its experimental state. The next insiders (Monday) will reflect this change.

**Comment by 4086606** [comment]
Could extension languages be allowed to "claim" or associate with a language model? Possibly through a contributes.languages field

**Comment by TylerLeonhardt** [comment]
Guesslang merged in the big PR that adds 24 more languages I think 🎉 I'll get that updated in vscode today. Yay JSON support!

**Comment by TylerLeonhardt** [comment]
> TensorFlow assigns a language ID, maybe erlang in this case, but how can we be sure the contributed extension language ID is erlang? Will this be done by staff when building the model?

Today, the model uses file extensions. Here's an example:
```
[
  { languageId: 'ts', confidence: 0.2529011368751526 },
  { languageId: 'js', confidence: 0.14790849387645721 },
  { languageId: 'rs', confidence: 0.14766104519367218 },
  { languageId: 'pm', confidence: 0.06276547163724899 },
  { languageId: 'dart', confidence: 0.051522497087717056 },
  { languageId: 'lua', confidence: 0.03615597262978554 },
  { languageId: 'go', confidence: 0.03003396838903427 },
  { languageId: 'cpp', confidence: 0.023810705170035362 },
  { languageId: 'html', confidence: 0.02115756832063198 },
  { languageId: 'md', confidence: 0.020726678892970085 },
  { languageId: 'c', confidence: 0.019782979041337967 },
  { languageId: 'prolog', confidence: 0.017745675519108772 },
  { languageId: 'php', confidence: 0.0135722691193223 },
  { languageId: 'pas', confidence: 0.013270312920212746 },
  { languageId: 'css', confidence: 0.01290779747068882 },
  { languageId: 'groovy', confidence: 0.012817374430596828 },
  { languageId: 'cs', confidence: 0.012597348541021347 },
  { languageId: 'ps1', confidence: 0.010909260250627995 },
  { languageId: 'swift', confidence: 0.008903573267161846 },
  { languageId: 'erl', confidence: 0.005858836695551872 },
  { languageId: 'sql', confidence: 0.005721408873796463 },
  { languageId: 'jl', confidence: 0.005521455779671669 },
  { languageId: 'mm', confidence: 0.005181039683520794 },
  { languageId: 'scala', confidence: 0.005087847821414471 },
  { languageId: 'r', confidence: 0.005077976267784834 },
  { languageId: 'sh', confidence: 0.004781287629157305 },
  { languageId: 'coffee', confidence: 0.004568156786262989 },
  { languageId: 'kt', confidence: 0.0038636510726064444 },
  { languageId: 'dm', confidence: 0.003127841744571924 },
  { languageId: 'rb', confidence: 0.0030808558221906424 },
  { languageId: 'py', confidence: 0.002654825570061803 },
  { languageId: 'ml', confidence: 0.002621284918859601 },
  { languageId: 'f90', confidence: 0.002495664870366454 },
  { languageId: 'java', confidence: 0.002417698036879301 },
  { languageId: 'bat', confidence: 0.0023386385291814804 },
  { languageId: 'v', confidence: 0.002330552088096738 },
  { languageId: 'tex', confidence: 0.0023196381516754627 },
  { languageId: 'ini', confidence: 0.001619357499293983 },
  { languageId: 'matlab', confidence: 0.0016029247781261802 },
  { languageId: 'hs', confidence: 0.0015097145223990083 },
  { languageId: 'asm', confidence: 0.0014708234230056405 },
  { languageId: 'yaml', confidence: 0.0013949687127023935 },
  { languageId: 'xml', confidence: 0.001296907546930015 },
  { languageId: 'vba', confidence: 0.0010232512140646577 },
  { languageId: 'ex', confidence: 0.0008915567887015641 },
  { languageId: 'csv', confidence: 0.0008855632040649652 },
  { languageId: 'makefile', confidence: 0.0006489697261713445 },
  { languageId: 'json', confidence: 0.0004926725523546338 },
  { languageId: 'lisp', confidence: 0.0003874319954775274 },
  { languageId: 'toml', confidence: 0.0003317541559226811 },
  { languageId: 'clj', confidence: 0.00010257939720759168 },
  { languageId: 'cbl', confidence: 0.00006274027691688389 },
  { languageId: 'dockerfile', confidence: 0.000046700715756742284 },
  { languageId: 'cmake', confidence: 0.00003329188984935172 }
]
```

we take that `languageId` and run it through VS Code's internal API to map that "model language id" to VS Code's language mode:

```ts
// TODO: see if there's a better way to do this.
const vscodeLanguageId = this._modeService.getModeIdByFilepathOrFirstLine(URI.file(`file.${possibleLanguages.shift()!.languageId}`));
```
> I included the TODO so to be clear that I'm not sure this is the best way to do this... :) 

There could be a situation where two languages share the same file extension. This is something that needs to be handled...

**Comment by TylerLeonhardt** [comment]
I've now updated the model to what is in the master branch in @yoeo's guesslang repo 🎉 tomorrow's insiders will support JSON, YAML, and more!

(also I debounced the event so perf should be way better now)

**Comment by TylerLeonhardt** [comment]
I will turn this feature on by default first thing next week.

**Comment by TylerLeonhardt** [comment]
This has been on by default for a while and no one has yelled. Also I:

* got the model size down to 1mb
* added an "Auto-detect" option in the Notebook language picker

**Comment by TylerLeonhardt** [comment]
I think this issue is basically done at this point so I'm going to close it.
